### ISSUE TITLE
SignalProducer on background scheduler not sending value events.

### ISSUE THREAD

**Issue by ABeard89** [ReactiveCocoa/ReactiveSwift#587]
I'm currently developing an app, and the API server isn't ready yet, so I'm trying to mock the API calls.

I'm having trouble, and I'd be very grateful for some help.

Here's basically what the API Manager looks like:

    class ApiManager {
        static let shared = ApiManager()
        private let scheduler = QueueScheduler(qos: .background, name: "MyProject.ApiManager")
        
        func getDataA() -> SignalProducer<[MyData], NoError> {
            return self.mockGetData(count: 1, delay: 1)
        }
        func getDataB() -> SignalProducer<[MyData], NoError> {
            return self.mockGetData(count: 2, delay: 1)
        }
        func getDataC() -> SignalProducer<[MyData], NoError> {
            return self.mockGetData(count: 3, delay: 1)
        }
        
        private func mockGetData(count: Int, delay: TimeInterval) -> SignalProducer<[MyData], NoError> {
            return SignalProducer<[MyData], NoError> { observer, lifetime in
                var dataItems = [MyData]()
                
                for _ in 1 ... count {
                    let myData = DataCache.shared.randomDataItem() // Simply gets a random item from an array
                    dataItems.append(myData)
                }
                
                observer.send(value: dataItems)
                observer.sendCompleted()
            }
            .delay(delay, on: self.scheduler)
        }
    }

In another place, I attempt to get one array of items from each A, B, and C.
That looks something like this.

    // Inside a class:
    
    // self.shouldGetDataProperty is of type MutableProperty<()>
    self.shouldGetDataProperty.signal
        .flatMap(.latest) { () -> SignalProducer<([MyData], [MyData], [MyData]), NoError> in
            return SignalProducer.zip(
                ApiManager.shared.getDataA(), // Sometimes terminates without values
                ApiManager.shared.getDataB(), // Sometimes terminates without values
                ApiManager.shared.getDataC() // Sometimes terminates without values
            )
        }
        .map { dataA, dataB, dataC in
            // Sometimes, this closure is not called.
            
            // Map and continue processing...
        }
    
My problem, as I wrote in the comments, is that sometimes (about 5%) one or more of the the zipped `SignalProducer`s doesn't return values, canceling the zip.

If I don't delay the signal producers in `ApiManager`, then this never happens. So, I suspect I'm doing something wrong with the `QueueScheduler `.

I tried logging events on each `SignalProducer` individually. When I do that, I typically get an event log like this:

> A: Starting
> A: Started
> B: Starting
> B: Started
> C: Starting
> C: Started
> A: Value [...]
> A: Completed
> A: Terminated
> A: Disposed
> B: Value [...]
> B: Completed
> B: Terminated
> B: Disposed
> C: Completed
> C: Terminated
> C: Disposed

In that example, `SignalProducer` C didn't send a `Value` event, despite that fact that a value had been sent to the `Observer` in the `SignalProducer`'s initialization closure.

What could I have done wrong to make the `SignalProducer` behave this way?

**Comment by mdiep** [comment]
Have you tried reducing your problem to a shareable, reproducible sample? I don't see that you've done anything wrong, which makes me think the problem may be with the code you haven't shared.

**Comment by ABeard89** [comment]
I’ll try to reduce it to something reproducible.

As for unshared code, is there any area of the surrounding code that you think might be suspect? For example, the method for grabbing a random data item? Or perhaps what happens in the mapping closure that isn’t even called?

The fact that it only happens sometimes (and so rarely), screams “race condition” to me, but I’m not sure where to go from there. Also, one more fact that makes me think race condition is that this problem never happens on the simulator. It only ever happens when running on a real device.

**Comment by mdiep** [comment]
That certainly sounds strange. I'm unsure what to think here. Hopefully a reproducible sample will make it more obvious.

**Comment by ABeard89** [comment]
Possibly a dumb question: could the iOS version play any part in this?

The reason I ask is the problem device is my company's iPhoneX running iOS 11.1.

While I can't reproduce the problem using the 11.1 simulator, I also can't reproduce the problem on an iPhone 7 running 11.2.1.

So, as far as I can tell, it's either:
1. A bug in the very low levels of iOS 11.1
or
2. A race condition that doesn't appear on the iPhone7 because it isn't as fast as the iPhoneX

Before I ask for permission to update the iPhoneX's software, is an iOS 11.1 bug a possibility?

**Comment by andersio** [comment]
Have you observed the `Signal` you composed?

**Comment by ABeard89** [comment]
Yes, it’s bound to a `MutableProperty` stored as an i-var in a view model. The view controller has a strong reference to the view model.

The closure that isn’t called simply maps the zipped tuple to a struct for the `MutableProperty`.

**Comment by ABeard89** [comment]
There’s one other fact that I think is interesting. This very signal is triggered by two separate events:
1. On a pull to refresh event in a collection view
2. On `viewDidAppear` for the initial load

I’ve discovered that case 1 never ever fails. 2 is the problem case. Both cases simply call a method on the view model to start that signal. Since it’s the same view model method, I can’t imagine how the two cases could behave differently.

**Comment by ABeard89** [comment]
What makes this so bizarre, is I can't reproduce the problem on anything but an actual iPhoneX running iOS11. This problem doesn't appear on any simulator or any other device or iOS version.

And since the problem only appears on the initial load of a main view controller, I basically have to just kill the app and restart it (or build/run with Xcode) on the device over and over again, which is a pain in the neck. And even then, it only appears about 5% of the time, so I have to restart the app about 20 times before the problem appears again.

It's just so weird. The composed signal never dies, but the inner `SignalProducer`'s `Observer` fails to send the `Value`?

**Comment by ABeard89** [comment]
I've tried to create a sample app to demonstrate the problem with a simple label in a view (rather than a collection view, like my app uses), but that app never shows the problem. So, I'm kinda stuck. I'm not sure where to go from here.

**Comment by ABeard89** [comment]
I actually did make some progress on this. I made a sample app to help reproduce the problem. It uses a couple coordinator objects, like my app does. I left out the view models to help keep it simple and instead just did that stuff in the view controller itself.

I put it on my github account: https://github.com/ABeard89/RASIssueTest

There's not much too it, just a simple label on the screen the says "Loading..." and then a composed signal like above changes it to "Loaded!".

When the zipped signal fails, the screen just says "Loading..." and will never switch to "Loaded!"

The bug appearance rate is very low on my iPhone7–less than 2%. But on the iPhoneX, it's higher. I tested repeatedly starting it with the build/run with Xcode method, and it shows up around every 10-20th time, rate of about 5-10%.

I really hope it shows up with you. The appearance rate is low, but I swear to god it happens over here, and it's even shown up on my iPhone7 a couple times. Even if it doesn't happen with you, maybe you can tell me what I've done wrong.

**Comment by andersio** [comment]
Sorry for the late response. I will look into it before the weekend, since the way you use it does seem normal to me.

**Comment by ABeard89** [comment]
No big rush on my end. Let me know if you want me to do anything.

**Comment by andersio** [comment]
I did reproduce it with an iPhone 8. Investigating. 😞

**Comment by andersio** [comment]
Unfortunately my hands are bit tight right in the past few weeks. I plan to to pick up the trails in the weekend. 😞

**Comment by andersio** [comment]
The weird thing is, now that I have time, I wasn't able to reproduce it on 11.2.5 like I did back in Jan.

**Comment by andersio** [comment]
I guess I would try to turn the example app into a stress test and see how it performs.

**Comment by andersio** [comment]
[The stress testing version](https://github.com/andersio/RASIssueTest/blob/anders/test/RAS%20Issue%20Test/Issue/IssueViewController.swift) which stops when it detects a discrepancy. Do you still happen to have an iOS 11.1 device at your hand?

**Comment by andersio** [comment]
Hmm, I did manage to narrow it down & it happens only when `QueueScheduler` is involved in the API manager.

**Comment by ABeard89** [comment]
I don't have any 11.1 devices anymore. We updates ours to 11.2. Although, I think iOS version was a red herring, since I've seen the error on iOS 9, 10, and 11 (even on 11.2) and on several different devices.

**Comment by ABeard89** [comment]
As a side note, thanks for working on this. I'm sure you do this in your spare time.

I've since gotten a full API for my app, so I don't depend on a fix for the issue, but I'm curious to see where this leads.

**Comment by ABeard89** [comment]
The weird thing about reproducing it was it seemed to go through phases of happening very frequently to phases of almost never happening at all. But of course, maybe that's just statistically natural.

**Comment by ABeard89** [comment]
I noticed that when I removed the delay via `QueueScheduler` in the API manager I was never able to reproduce it. So, I was beginning to suspect that it was some kind of dispatch issue, which is far above my ability to debug.

**Comment by ABeard89** [comment]
A strange thing happened just now. It makes me think that this problem might not be in ReactiveSwift, but rather somewhere much deeper, like GCD or the Swift compiler or Xcode.

Here's what I saw that makes me think this:

**TL;DR:** Closure passed to `UIKit` started not being called, but then started being called again, for no reason.

I was working on another part of my app where I'm observing signals from a view model.
Here's a snippet of that

    private func listen(to viewModel: MenuViewModel) {    
        viewModel.outputs.openPurchaseHistoryScreen
            .take(duringLifetimeOf: self)
            .observeValues { [unowned self] in
                self.closeMenu() { [unowned self] in
                    self.delegate?.openPurchaseHistory()
                }
        }
        
        viewModel.outputs.openCouponScreen
            .take(duringLifetimeOf: self)
            .observeValues { [unowned self] url in
                self.closeMenu() { [unowned self] in
                    self.delegate?.openCoupon(url)
                }
        }
        
        viewModel.outputs.openViewHistory
            .take(duringLifetimeOf: self)
            .observeValues { [unowned self] in
                self.closeMenu() { [unowned self] in
                    self.delegate?.openViewHistory()
                }
        }
    }

This section hasn't changed in ages. This is the exact code currently on the App Store.


The strange thing was that in my debug builds, suddenly the last closure calling `self.delegate?.openViewHistory()` simply wasn't being called.
`closeMenu` was called. The menu even properly closed, but the closure wasn't called.

I'm not sure why. I followed the closure all the way to `UIViewController.dismiss(animated:completion:)`. The closure existed as far as I could track it, but it simply wasn't called.

I tried cleaning the build folder, restarting Xcode, and building on another device; but nothing changed.

What did work was simply changing the order of the observations in the method. (e.g. `openViewHistory` before `openCouponScreen`)
Then every closure was called as expected, but when I put them back to their original order, the problem started happening again.

It simply looked like when the `openViewHistory` observation was last in the method, the closure passed to `closeMenu` just wasn't be called.


I tried several experiments to no effect, then tried building again with the original code, and then everything magically started working again!

The code from when it wasn't working and when it was working are exactly the same code. But something made it either get built differently or called differently, I'm not sure which.


Have you ever heard of anything like this? Do you think this might be related?

**Comment by ABeard89** [comment]
(I accidentally closed this just now. My mistake. Just ignore that.)

**Comment by andersio** [comment]
Sounds like a corrupted build. :s

**Comment by ABeard89** [comment]
Even after cleaning and building on different devices multiple times?

If so, thanks, Xcode... ugh

**Comment by RuiAAPeres** [comment]
Hello. 👋  Thanks for opening this issue. Due to inactivity, we will soft close the issue. If you feel that it should remain open, please let us know. 😄
